packages/core/core/src/Transformation.js

```diff
@@ -0,0 +1,499 @@
+  Array<InternalAsset>
+) => Promise<Array<InternalAsset> | null>;
+
+const BUFFER_LIMIT = 5000000; // 5mb
+
+export type TransformationOpts = {|
+  request: AssetRequest,
+  loadConfig: (ConfigRequest, NodeId) => Promise<Config>,
+  parentNodeId: NodeId,
+  options: ParcelOptions
+|};
+
+type ConfigMap = Map<PackageName, Config>;
+
+export default class Transformation {
+  request: AssetRequest;
+  loadConfig: ConfigRequest => Promise<Config>;
+  options: ParcelOptions;
+  cache: Cache;
+  envId: string;
+  impactfulOptionsId: string;
+
+  constructor({
+    request,
+    loadConfig,
+    parentNodeId,
+    options
+  }: TransformationOpts) {
+    this.request = request;
+    this.loadConfig = configRequest => loadConfig(configRequest, parentNodeId);
+    this.options = options;
+
+    // TODO: these options may not impact all transformations, let transformers decide if they care or not
+    let {minify, hot, scopeHoist} = this.options;
+    this.impactfulOptionsId = JSON.stringify({minify, hot, scopeHoist});
+
+    this.envId = JSON.stringify(this.request.env);
+  }
+
+  async run() {
+    report({
+      type: 'buildProgress',
+      phase: 'transforming',
+      request: this.request
+    });
+
+    this.cache = new Cache(this.options.cacheDir);
+
+    let asset = await this.loadAsset();
+    let pipeline = await this.loadPipeline(asset.filePath);
+    return this.runPipeline(pipeline, asset);
+  }
+
+  async loadAsset() {
+    let {filePath, env, code, sideEffects} = this.request;
+    let {content, size, hash} = await summarizeRequest(this.request);
+
+    return new InternalAsset({
+      // If the transformer request passed code rather than a filename,
+      // use a hash as the base for the id to ensure it is unique.
+      idBase: code != null ? hash : filePath,
+      filePath: filePath,
+      type: path.extname(filePath).slice(1),
+      cache: this.cache,
+      ast: null,
+      content,
+      hash,
+      env: env,
+      stats: {
+        time: 0,
+        size
+      },
+      sideEffects: sideEffects
+    });
+  }
+
+  async runPipeline(pipeline: Pipeline, initialAsset: InternalAsset) {
+    let initialType = initialAsset.type;
+    let initialCacheEntry = await this.readFromCache(
+      [initialAsset],
+      pipeline.configs
+    );
+
+    let assets = initialCacheEntry || (await pipeline.transform(initialAsset));
+    if (!initialCacheEntry) {
+      await this.writeToCache(assets, pipeline.configs);
+    }
+
+    let finalAssets: Array<InternalAsset> = [];
+    for (let asset of assets) {
+      let nextPipeline;
+      if (asset.type !== initialType) {
+        nextPipeline = await this.loadNextPipeline(
+          initialAsset.filePath,
+          asset.type,
+          pipeline
+        );
+      }
+
+      if (nextPipeline) {
+        let nextPipelineAssets = await this.runPipeline(nextPipeline, asset);
+        await Promise.all(
+          nextPipelineAssets.map(nextAsset => nextAsset.getCode())
```

Comment:
Why do we do this? This is going to force reading entire streams that could be filled with large binary data.

Reply:
I chose to cache the at the pipeline layer because I felt it was easier to follow how intermediate assets can get cached. Also now that the cache keys have all of the invalidation information, it means we don't have can still invalidate individual pipeline results. I get that it's not ideal to commit it and then read it right away again. We should probably have a way we can commit the result of a pipeline without the content on the assets while still keeping the content on the in memory assets.

Reply:
I still don't understand what effect this has. Doesn't look like you're using the results of getting the code right?

Also, this won't work with binary assets for example. We have separate methods for those: `getCode`, `getBuffer`, `getMap`, etc. and only the transformer plugin will know which to call.

Reply:
> I still don't understand what effect this has. Doesn't look like you're using the results of getting the code right?

I guess I was thinking that `Asset.commit()` cleared code/content from the asset. Is it only the serialization that does that?